node/basicnode/map_test.go

```diff
@@ -46,3 +51,277 @@ func BenchmarkSpec_Unmarshal_Map3StrInt(b *testing.B) {
 func BenchmarkSpec_Unmarshal_MapNStrMap3StrInt(b *testing.B) {
 	tests.BenchmarkSpec_Unmarshal_MapNStrMap3StrInt(b, basicnode.Prototype.Map)
 }
+
+// Test that the map builder cannot be assigned arbitrary values, and trying to
+// will result in a sensible error
+func TestMapAssignError(t *testing.T) {
+	b := basicnode.Prototype.Map.NewBuilder()
+	err := b.AssignBool(true)
+	errExpect := `func called on wrong kind: "AssignBool" called on a map node \(kind: map\), but only makes sense on bool`
+	qt.Check(t, err, qt.ErrorMatches, errExpect)
+
+	err = b.AssignInt(3)
+	errExpect = `func called on wrong kind: "AssignInt" called on a map node \(kind: map\), but only makes sense on int`
+	qt.Check(t, err, qt.ErrorMatches, errExpect)
+
+	err = b.AssignFloat(5.7)
+	errExpect = `func called on wrong kind: "AssignFloat" called on a map node \(kind: map\), but only makes sense on float`
+	qt.Check(t, err, qt.ErrorMatches, errExpect)
+
+	err = b.AssignString("hi")
+	errExpect = `func called on wrong kind: "AssignString" called on a map node \(kind: map\), but only makes sense on string`
+	qt.Check(t, err, qt.ErrorMatches, errExpect)
+
+	err = b.AssignNode(basicnode.NewInt(3))
+	errExpect = `func called on wrong kind: "AssignNode" called on a map node \(kind: int\), but only makes sense on map`
+	qt.Check(t, err, qt.ErrorMatches, errExpect)
+
+	// TODO(dustmop): BeginList, AssignNull, AssignBytes, AssignLink
+}
+
+// Test that the map builder can create map nodes, and AssignNode will copy
+// such a node, and lookup methods on that node will work correctly
+func TestMapBuilder(t *testing.T) {
+	b := basicnode.Prototype.Map.NewBuilder()
+
+	// construct a map of three keys, using the MapBuilder
+	ma, err := b.BeginMap(3)
+	if err != nil {
+		t.Fatal(err)
+	}
+	a := ma.AssembleKey()
+	a.AssignString("cat")
+	a = ma.AssembleValue()
+	a.AssignString("meow")
```

Comment:
not strictly necessary here since we have coverage already, but you could also be checking the intermediate errors instead of dropping them on the floor; it can be a one-liner for all the `Assign*`s:

```suggestion
	qt.Check(t, a.AssignString("meow"), qt.IsNil)
```

Reply:
If it's alright, I'd prefer to keep this simple for readability. It's easy enough to convince oneself that this statement (and others like it) won't fail, but if they somehow did, the test will certainly not pass.